## Project Pt 3

### 1. Introduction
This paper completed a meta-analysis examining the differences in microbiomes of corals with and without stony coral tissue loss disease (sctld). In an effort to reduce biases from varying sampling and bioinformatic methods, they compiled datasets from 16 different studies to identify microbial taxa that may be associated with the disease since the cause of sctld is still unknown. 

### 2. Project Pt 2

#### 2.1 Contents of Project pt 2
Section 1: Data Set 1 from Project Pt1: https://www.nature.com/articles/s43705-023-00220-0
- Data sets acquired from 16 studies. **2425** total coral samples (observations). samples collected either from the field or from lab-based experiments and sequenced  
- Features (>1000) - taxonomic units (ASVs) that describe how many counts of those ASVs were in each coral sample (the observations)
- Metadata about observations - sample type (tissue, mucus, mucus/tissue combo), disease state, geographic location, coral species, aquaria vs natural environment. They also had other data sets of water and sediment samples with ASVs as features. 

Research Qs:
- the cause of stony coral tissue loss disease is unknown but is hypothesized to be from a microbial pathogen - what bacteria are associated with the disease? How do these associated taxa vary across environments and coral species?
    - they theorized that a lack of consensus across studies investigating bacteria associated with sctld is from variability in both sample type (location, coral species, field vs aquaria) and bionformatic analysis. They collected data from multiple studies to analysis them using a single bioinformatic pipeline to limit this variability and investigate similar questions by a meta-analysis using the compiled data.
- you can also ask how coral microbiomes differ when raised in aquaria vs found in their natural environments. you can also examine how microbiomes of the same species differ across locations and disease states. **comment correction: The data set is useful because it includes microbiome samples for corals raised in aquaria vs those sampled from the field. Since the corals sampled have varying disease states, you can just reframe the question to be about coral microbiomes in different environments instead of focusing on disease state. There is also plenty of samples from the same species which is why this data would be useful in examning microbiomes across disease state and/or location within the same host species.** 
- they also examined the microbial communities in the water and sediment surrounding some of the coral samples to try to identify potential pathogens that are transmissable 


Figure 1: 
![image.png](attachment:aa9d2e13-d310-4a79-b5d4-d43e7879f975.png)

Section 2 - **Figure 1** Answers:
- x-axis: Log fold change & relative abundance
- y-axis: taxa (genus, order)
- A) the data points show the log fold changes in taxonomic groups when comparing disease lesion (diseased tissue sample) to apparently healthy. So, anything showing positive log-fold change means that taxa was more abundant in Disease Lesion, and anything negative was less abundant in Disease Lesion
- B) the relative abundances of the taxa groups within the sample types. **comment correction: the relative abundance is calculating the fraction of sample type compared to the rest. the number of samples within a sample type (seawater, mucus, slury, etc) divided by the total number of samples - and the total number of samples is set to 1**
- this figure illustrates how bacterial taxa significantly change between healthy and diseased corals, supporting their hypothesis that certain taxa are associated with the disease. B shows how some of these abundances vary across environment or parts of the coral (mucus vs tissue) to support their hypothesis that microbiome changes and associated bacteria may depend on individual hosts. It also shows how disease-associated microbes may be present in the water and sediment, providing evidence for their hypothesis that this disease is transmissable.
- Conclusions: Certain bacterial groups are associated with the disease. Multiple genera in Rhodobacterales and Peptostreptococcales increased significantly in diseased samples indicating it may be pathogenic or just associated with the disease state. Some bacteria associated with the disease may be present in the surrounding seawater and sediment, meaning it may be transmissable. 

Figure 2: ![image.png](attachment:f53cb107-74cb-4be2-ad39-1a5741094d38.png)

Section 2 - **Figure 2** Answers:
- x-axis: Log fold change & relative abundance
- y-axis: taxa (genus, order)
- A) the data points show the log fold changes in taxonomic groups when comparing disease unaffected (unaffected tissue sample from a diseased coral colony) to apparently healthy. So, anything showing positive log-fold change means that taxa was more abundant in Disease Unaffected, and anything negative was less abundant in Disease Unaffected
- B) the relative abundances of the taxa groups within the sample types

- this figure also illustrates how bacterial taxa significantly change between healthy and diseased corals, supporting their hypothesis that certain taxa are associated with the disease. This is helpful to provide in addition to Figure 1 because it shows that the microbiome of the entire coral colony changes after being introduced to the disease, even if the disease symptoms haven't spread through the whole coral colony. B shows how some of these abundances vary across environment or parts of the coral (mucus vs tissue) to support their hypothesis that microbiome changes and associated bacteria may depend on individual hosts. It also shows how disease-associated microbes may be present in the water and sediment, providing evidence for their hypothesis that this disease is transmissable.

- Conclusions: Both figures together show which bacteria may be associated with healthy corals or the disease. Multiple genera in Rhodobacterales decreased significantly in DU group but Flavobacteriales increased. 

Another figure that would be helpful would be the same type of figure showing the significant changes in taxa between Disease Unaffected (DU) and Disease Lesion (DL). This would allow you to easily compare all the taxa between all the groups - especially between the disease lesion and disease unaffected. this direct comparison is currently not shown in a figure in the paper. 

Section 3: 
Data table 1: taxonomy.tsv 
- tab deliminated data table - opened in excel 
- 147455 rows & 3 columns. The columns are the feature IDs (which is just the sequence sample identifier), Taxonomy (the identified microbial taxa from the sequence: kingdom - species), and Concensus (percentage that the sequence matched the identified taxa group in the taxonomy database it was matched against). The feature ID's are the observations and the taxonomy and consensus columns are the features. 
- This data is important because it matches each sequence to a microbial taxa. This enables comparing the taxa between samples which is important to answer the paper's questions. 

Data table 2: Feature-table.biom
- biom table format - downloaded QIIME2 (bioinformatic software package) and accessed through command line: converted biom format to txt file and opened as tab deliminated table in excel
- 119456 rows, 2426 columns. rows are feature IDs (match the ones in data table 1), columns are samples (observations). File contains number of feature IDs (sequences of microbial taxa) in each coral sample
- this is essential to the paper because knowing the amount of each taxa in the samples allows comparisons between sample groups, so you can identify bacteria associated with diseased coral samples by seeing which taxa are more abundant in diseased samples vs healthy samples.

Data table 3: SCTLD_meta_analysis_metadata.csv
- comma deliminated table - opened in excel 
- 2444 rows, 68 columns. rows are sample IDs the columns contain metadata about the samples including the study the data came from, the coral species, the disease state, location of coral. 
- The file is important to the analysis because it allows you to match the sample ID to the other data tables and contains important information about the samples. Using this in combination with the other data tables, they were able to sort samples into disease state or environment or sample type to compare microbial abundances to answer their research question.

#### 2.2 Project Pt 2
- I added why the data set on hand would be useful for other potential research questions and added more information on what relative abundance (one of the axes) is in figure 1. 

### 3. Data Cleaning and Exploration

#### 3.1 Load and clean data
Link to paper & data:
https://www.nature.com/articles/s43705-023-00220-0#data-availability

##### Steps to obtain data:
1. Click link to github with data 
https://github.com/srosales712/SCTLD_microbiome_meta_analysis

2. Download SCTLD_meta_analysis_metadata.csv and save into ds4bio directory

3. Go to Qiime_output folder and download feature-table.biom & taxonomy.tsv and save into ds4bio directory

4. Complete code below to convert the "biom" format into a tsv file

In [ ]:
#install biom-format via command line in terminal 
pip install biom-format
#convert the biom table to a tsv file 
biom convert -i feature-table.biom -o feature-table.tsv --to-tsv

5. Open the feature-table.tsv file in excel (select "tab-deliminated" when prompted)

6. Open the taxonomy.tsv file in excel the same way 

7. Open the SCTLD_meta_analysis_metadata.csv in excel

##### Loading Data

In [3]:
#import packages 
import pandas as pd
import numpy as np

#load in the metadata from csv as a dataframe. index_col=0 to make row names the first column
metadata = pd.read_csv("SCTLD_meta_analysis_metadata.csv", index_col = 0)
metadata.head(3)

SampleID     BioSample Run SRA_accession      Lab  \
sample-id                                                     
66NAT        66NAT  SAMN14322012  12    SRS6274849  P_Thome   
68NAT        68NAT  SAMN14322013  12    SRS6274852  P_Thome   
69NAT        69NAT  SAMN14322014  12    SRS6274848  P_Thome   

                        Experiment                    Alias    Study  \
sample-id                                                              
66NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   
68NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   
69NAT      Pseudodiploria_strigosa  Patriciathome_Pstrigosa  P_Thome   

          Collection_ID Species_counts  ... primer_names      forwardPrimer  \
sample-id                               ...                                   
66NAT             66NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   
68NAT             68NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   
69NAT             69NAT           High  ...    341F_805R  CCTACGGGNGGCWGCAG   

                   ReversePrimer  \
sample-id                          
66NAT      GACTACHVGGGTATCTAATCC   
68NAT      GACTACHVGGGTATCTAATCC   
69NAT      GACTACHVGGGTATCTAATCC   

                                        linkerPrimerSequence  platform  \
sample-id                                                                
66NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   
68NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   
69NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   

          platformChemistry sequencing_facility   barcodes basepairs notes  
sample-id                                                                   
66NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  
68NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  
69NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  

[3 rows x 69 columns]

##### Data Cleaning:

In [4]:
#from viewing the meta dataframe: index and 1st column are duplicates, so remove first column
meta=metadata.drop("SampleID",axis=1)
meta.head(3)

BioSample Run SRA_accession      Lab               Experiment  \
sample-id                                                                     
66NAT      SAMN14322012  12    SRS6274849  P_Thome  Pseudodiploria_strigosa   
68NAT      SAMN14322013  12    SRS6274852  P_Thome  Pseudodiploria_strigosa   
69NAT      SAMN14322014  12    SRS6274848  P_Thome  Pseudodiploria_strigosa   

                             Alias    Study Collection_ID Species_counts  \
sample-id                                                                  
66NAT      Patriciathome_Pstrigosa  P_Thome         66NAT           High   
68NAT      Patriciathome_Pstrigosa  P_Thome         68NAT           High   
69NAT      Patriciathome_Pstrigosa  P_Thome         69NAT           High   

          collection_time  ... primer_names      forwardPrimer  \
sample-id                  ...                                   
66NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   
68NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   
69NAT                 NaN  ...    341F_805R  CCTACGGGNGGCWGCAG   

                   ReversePrimer  \
sample-id                          
66NAT      GACTACHVGGGTATCTAATCC   
68NAT      GACTACHVGGGTATCTAATCC   
69NAT      GACTACHVGGGTATCTAATCC   

                                        linkerPrimerSequence  platform  \
sample-id                                                                
66NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   
68NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   
69NAT      TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG/GTCTCGTGGGCT...  Illumina   

          platformChemistry sequencing_facility   barcodes basepairs notes  
sample-id                                                                   
66NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  
68NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  
69NAT                 MiSeq    CINVESTAV_Merida  NexteraXT     250.0   NaN  

[3 rows x 68 columns]

##### Loading Data

In [5]:
#load in the taxonomy from tsv as a dataframe. index_col=0 to make row names the first column, and it is tab seperated instead of comma
taxonomy = pd.read_csv("taxonomy.tsv",sep='\t', index_col = 0)
taxonomy.head(3)

Taxon  \
Feature ID                                                                            
f4f2731649deea1773042cf5d18a0a33  d__Eukaryota; p__Dinoflagellata; c__Dinophycea...   
dc99cb7f143b60159d9626ae5be10656                                         Unassigned   
82eb6703b24ac1c5e2e7384217cb6747  d__Bacteria; p__Bacteroidota; c__Bacteroidia; ...   

                                  Consensus  
Feature ID                                   
f4f2731649deea1773042cf5d18a0a33   1.000000  
dc99cb7f143b60159d9626ae5be10656   1.000000  
82eb6703b24ac1c5e2e7384217cb6747   0.666667

##### Data Cleaning:

In [16]:
#separate the different phylogenetic classifications (in the Taxon column) into different columns by the deliminater ;. expand=True makes it expand the results into different columns  
#https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html
taxonomy_levels=taxonomy['Taxon'].str.split(';', expand=True)

#remove the prefixes in the classifications in all taxa columns and replace with an empty string. regex=True argument means the replacement value will be a string 
#https://stackoverflow.com/questions/26837998/pandas-replace-nan-with-blank-empty-string
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html
taxonomy_levels2 = taxonomy_levels.replace('[a-z]__', '', regex=True)

#rename columns to correspond to taxonomic levels 
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html
taxonomy_levels_renamed=taxonomy_levels2.rename(columns={0: "Domain", 1: "Phylum", 2: "Class", 3:"Order", 4:"Family", 5: "Genus", 6: "Species"})

#concatenate renamed taxa dataframe to original dataframe (axis=1 to concat along columns)
taxa = pd.concat([taxonomy, taxonomy_levels_renamed], axis=1)
#remove original "Taxon" column
#https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html
taxa = taxa.drop('Taxon', axis=1)
taxa.head(3)

Consensus      Domain           Phylum  \
Feature ID                                                                 
f4f2731649deea1773042cf5d18a0a33   1.000000   Eukaryota   Dinoflagellata   
dc99cb7f143b60159d9626ae5be10656   1.000000  Unassigned             None   
82eb6703b24ac1c5e2e7384217cb6747   0.666667    Bacteria     Bacteroidota   

                                         Class               Order  \
Feature ID                                                           
f4f2731649deea1773042cf5d18a0a33   Dinophyceae   Gymnodiniphycidae   
dc99cb7f143b60159d9626ae5be10656          None                None   
82eb6703b24ac1c5e2e7384217cb6747   Bacteroidia    Flavobacteriales   

                                              Family              Genus  \
Feature ID                                                                
f4f2731649deea1773042cf5d18a0a33         Suessiaceae       Symbiodinium   
dc99cb7f143b60159d9626ae5be10656                None               None   
82eb6703b24ac1c5e2e7384217cb6747   Flavobacteriaceae   NS5_marine_group   

                                                Species  
Feature ID                                               
f4f2731649deea1773042cf5d18a0a33       Symbiodinium_sp.  
dc99cb7f143b60159d9626ae5be10656                   None  
82eb6703b24ac1c5e2e7384217cb6747   uncultured_bacterium

##### Loading Data

In [75]:
#unzip feature-table.tsv.gz in terminal (had to zip to upload to github)
#gunzip feature-table.tsv.gz 
#load in the feature-table from tsv as a dataframe. index_col=0 to make row names the first column, and it is tab seperated instead of comma
counts = pd.read_csv("feature-table.tsv",sep='\t', skiprows=1,index_col = 0)
counts.head(3)

FileNotFoundError: [Errno 2] No such file or directory: 'feature-table.tsv'

#### 3.2 Describe Data Numerically 

In [62]:
meta.shape

(2444, 68)

- The meta dataframe has 2444 rows corresponding to 2444 samples/observations (from 16 different studies - each contributing a different number of samples) 
- and 68 columns which corresponds to 68 different features of metadata (including the study the samples came from, collection info, etc.)

In [63]:
counts.shape

(119456, 2425)

- Counts dataframe has 119456 rows which are the features. Each corresponds to a unique OTU (Operational Taxonomic Unit) which is just different microbial groups 
- and 2425 columns - correspond to 2425 samples/observations (one study was excluded because it was not focused on  SCTLD) which is why it differs from the metadata which includes all the studies. 

In [64]:
taxa.shape

(147455, 8)

- Taxa dataframe has 147455 rows which are the features. Each corresponds to a unique feature ID - the same as an OTU (I am not sure why the use different terms)
- and 8 columns which contain the features - The taxonomic classifications which span 7 rows and the consensus which is the confidence value that each feature id matches the sequence. 

###### "Describe" function expectations:
- meta - feature 'colony_percent_diseased'
    - i think the average will be ~50%. I think about half of the observations will have that feature and it will vary from 0-100% so it will average around 50%
- meta - feature 'colony_max_diameter'
    - i think most observations have this feature except some studies that did not measure this. I would estimate the average will be around 100cm. 

In [70]:
meta['colony_percent_diseased'].describe()

count     1577
unique      30
top        TBD
freq       788
Name: colony_percent_diseased, dtype: object

- I did not realize that not all the entries were numerical values, so the describe function did not have an output that I expected
- there were many "TBD" entries that were unexpected, and because of all the string values it did not give the average value

In [71]:
meta['colony_max_diameter'].describe()

count     1134
unique      71
top        TBD
freq       439
Name: colony_max_diameter, dtype: object

- Similary to above, I did not realize that not all the entries were numerical values, so the describe function did not have an output that I expected
- there were many "TBD" entries that were unexpected, and because of all the string values it did not give the average value. Only 439 samples had measured diameters which was less than I expected. 

In [72]:
counts.head(5)

100tiss_S31  102tiss_S3  104tiss_S32  \
#OTU ID                                                                  
3c025d6a19302802d61600c5a47be873          0.0         0.0          0.0   
15632e3b84cd5b1e8146f6a3ddc9bc87          0.0         0.0          0.0   
53cfc5c67e76077d5cea348037152f9f          0.0         0.0          0.0   
70398316b6aa3a3411bb38f93f3f84d5          0.0         0.0          0.0   
b444ef705ef2a760ff3bfec26c6fe3b7          0.0         0.0          0.0   

                                  168eco_S8  171eco_S9  172eco_S10  \
#OTU ID                                                              
3c025d6a19302802d61600c5a47be873        0.0        0.0         0.0   
15632e3b84cd5b1e8146f6a3ddc9bc87        0.0        0.0         0.0   
53cfc5c67e76077d5cea348037152f9f        0.0        0.0         0.0   
70398316b6aa3a3411bb38f93f3f84d5        0.0        0.0         0.0   
b444ef705ef2a760ff3bfec26c6fe3b7        0.0        0.0         0.0   

                                  173eco_S11  175eco_S12  181eco_S14  \
#OTU ID                                                                
3c025d6a19302802d61600c5a47be873         0.0         0.0         0.0   
15632e3b84cd5b1e8146f6a3ddc9bc87         0.0         0.0         0.0   
53cfc5c67e76077d5cea348037152f9f         0.0         0.0         0.0   
70398316b6aa3a3411bb38f93f3f84d5         0.0         0.0         0.0   
b444ef705ef2a760ff3bfec26c6fe3b7         0.0         0.0         0.0   

                                  183eco_S15  ...  W4Aug23  W4Aug28  W4Oct2  \
#OTU ID                                       ...                             
3c025d6a19302802d61600c5a47be873         0.0  ...      0.0      0.0     0.0   
15632e3b84cd5b1e8146f6a3ddc9bc87         0.0  ...      0.0      0.0     0.0   
53cfc5c67e76077d5cea348037152f9f         0.0  ...      0.0      0.0     0.0   
70398316b6aa3a3411bb38f93f3f84d5         0.0  ...      0.0      0.0     0.0   
b444ef705ef2a760ff3bfec26c6fe3b7         0.0  ...      0.0      0.0     0.0   

                                  W4Sep9  W62Aug23  W62Aug26  W7Aug23  \
#OTU ID                                                                 
3c025d6a19302802d61600c5a47be873     0.0       0.0       0.0      0.0   
15632e3b84cd5b1e8146f6a3ddc9bc87     0.0       0.0       0.0     32.0   
53cfc5c67e76077d5cea348037152f9f     0.0       0.0       0.0      0.0   
70398316b6aa3a3411bb38f93f3f84d5     0.0       0.0       0.0      0.0   
b444ef705ef2a760ff3bfec26c6fe3b7     0.0       0.0       0.0      0.0   

                                  W7Aug28  W7Sep15  W7Sep9  
#OTU ID                                                     
3c025d6a19302802d61600c5a47be873      0.0      0.0     0.0  
15632e3b84cd5b1e8146f6a3ddc9bc87      0.0     20.0     0.0  
53cfc5c67e76077d5cea348037152f9f      0.0      0.0     0.0  
70398316b6aa3a3411bb38f93f3f84d5      0.0      0.0     0.0  
b444ef705ef2a760ff3bfec26c6fe3b7      0.0      0.0     0.0  

[5 rows x 2425 columns]

###### "Describe" function expectations:
- counts - feature (OTU ID) 3c025d6a19302802d61600c5a47be873
    - based on looking at the data it looks like there will be mostly 0s (meaning not many of the observations have this feature) and probably a few very high numbers, so I anticipate the average will be close to 0. 
- counts - feature (OTU ID) 15632e3b84cd5b1e8146f6a3ddc9bc87
    - same idea as above, not all of the observations have this feature so most of the values will be 0, and the average will be around 0. 

In [73]:
counts.loc['3c025d6a19302802d61600c5a47be873',:].describe()

count    2425.000000
mean        0.009485
std         0.467059
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max        23.000000
Name: 3c025d6a19302802d61600c5a47be873, dtype: float64

In [76]:
counts.loc['15632e3b84cd5b1e8146f6a3ddc9bc87',:].describe()

count    2425.000000
mean        0.146804
std         2.375596
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max        71.000000
Name: 15632e3b84cd5b1e8146f6a3ddc9bc87, dtype: float64

- as i expected, the mean was around 0. Though it was closer to 0 in the first feature. I did not expect the maxes to be so low on both of them. I thought they would be in the thousands. The std deviation also surprised me by being so low in the first one, but it makes since so many of the values were 0. 

- the mean was higher than i expected at about 80% meaning more observations than i expecetd had a 100% confidence value. I also expected the min to be lower, but they probably set the confidence value to be at least 60% when matching taxonomy which is why this is the min value. 

In [79]:
taxa['Domain'].describe()

count       147455
unique           4
top       Bacteria
freq        117916
Name: Domain, dtype: object

In [81]:
#print rows, columns of taxa
print(taxa.shape)
#obtain number of non-bacteria hits by subtracting the "freq" outcome of describe to the total number of observations obtained by the row number of shape
print(147455-117916)

(147455, 8)
29539


- 'Bacteria' was the top value as I had expected. 4 unique values also makes sense since there are 3 domains of life: eukaryota, bacteria, archaea and as I mentioned, some will say 'unassigned'. There were 29539 observations that were not bacteria though, and that is higher than i expected. 

#### 3.1 Visualizations